emission/storage/modifiable/builtin_model_storage.py

```diff
@@ -48,3 +52,34 @@ def get_current_model(self, key:str) -> Optional[Dict]:
             del first_entry["_id"]
             return first_entry 
 
+    def trim_model_entries(self, key:str):
+        """
+        :param: the metadata key for the entries, used to identify the model type
+        This function is called inside the model insertion function just before the 
+        model is inserted, to ensure older models are removed before inserting newer ones.
+
+        The flow of model insertion function calls is:
+        eamur.update_trip_model() -> eamums.save_model() -> esma.upsert_model() -> esma.trim_model_entries()
+        """
+        current_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
+        logging.debug("Before trimming, model count for user %s = %s" % (self.user_id, current_model_count))
+        find_query = {"user_id": self.user_id, "metadata.key": key}
+        result_it = edb.get_model_db().find(find_query).sort("metadata.write_ts", -1)
+        result_list = list(result_it)
+        maximum_stored_model_count = eamtc.get_maximum_stored_model_count()
+        if current_model_count >= maximum_stored_model_count:
+            # Specify the last or minimum timestamp of Kth model entry
+            write_ts_limit = result_list[maximum_stored_model_count - 1]['metadata']['write_ts']
+            logging.debug(f"Write ts limit = {write_ts_limit}")
+            filter_clause = {
+                "user_id" : self.user_id,
+                "metadata.key" : key,
+                "metadata.write_ts" : { "$lte" : write_ts_limit }
+            }
+            models_to_delete = edb.get_model_db().delete_many(filter_clause)
+            if models_to_delete.deleted_count > 0:
+                logging.debug(f"{models_to_delete.deleted_count} documents deleted successfully\n")
+            else:
+                logging.debug("No documents found or none deleted\n")
+        new_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
+        logging.debug("After trimming, model count for user %s = %s" % (self.user_id, new_model_count))
```

Comment:
Or maybe this is where we need the `old -> new` log. In that case, I am not sure what the previous log adds.

Reply:
Yes, it makes more sense to add the change in model count here.
I'll explain the need for the logging statement present at the beginning in `upsert_model()`.
To clarify the three different states where the logging statements are present, say we have 3 models currently.

Two sets of changes to model count are there: a) trimming models b) inserting new model

1. Current count before **trimming** models = 3
```
# In trim_model_entries()
current_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
logging.debug("Before trimming, model count for user %s = %s" % (self.user_id, current_model_count))
```

2. New count after **trimming** models = 2
```
# In trim_model_entries()
new_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
logging.debug("After trimming, model count for user %s = %s" % (self.user_id, new_model_count))
```

3. New count after **inserting** models = 3
```
# In upsert_model()
new_model_count = edb.get_model_db().count_documents({"user_id": self.user_id})
logging.debug("New model count for user %s = %s" % (self.user_id, new_model_count))
```